doc/design/python-api-1.md

```diff
@@ -0,0 +1,187 @@
+Designs for an improved Python API
+==================================
+
+* New base classes:
+    * `Dandiset`
+        * `identifier: str`
+        * `created: datetime`
+        * `modified: datetime`
+        * `version: Optional[Version]`
+        * `get_all_assets() -> Iterator[Asset]`
+        * `get_assets_under_path(path: str) -> Iterator[Asset]` — Returns all assets whose paths are either under the given folder or equal the given filepath
```

Comment:
UX sugarings we should consider add on top
```
@assets -> get_assets()
@metadata -> get_metadata()
__iter__() -> .get_assets()
```
?  they might be tricky with subclassing though since IIRC we cannot just bind the property to get _assets in super class, since then subclasses .assets would us it instead of overloaded...   May be properties should be lazy (cached value) invocations over the methods?  That would give them a better purpose

Reply:
I'm not clear on what exactly you're proposing here.

Reply:
smth like

```python
@property
def assets(self):
     if self._assets is None:
            self._assets = self.get_assets()
     return self._assets

def __iter__(self):
     for asset in self.assets:
          yield asset
```

but we can think about all the sugaring later